rust/config/src/ffi.rs

```diff
@@ -0,0 +1,392 @@
+            return std::ptr::null_mut();
+        }
+    };
+
+    let config = match crate::loader::load_from_file(filename) {
+        Ok(mut docs) => docs.pop().unwrap(),
+        Err(err) => {
+            let error = match err {
+                LoaderError::NotAFile(_) => "Provided filename is not a file".to_string(),
+                LoaderError::YamlScanError { filename, source } => {
+                    let error = format!(
+                        "YAML scan error at line {}, col {}",
+                        source.marker().line(),
+                        source.marker().col()
+                    );
+
+                    // Include the filename if there is one, the error could originate
+                    // from an include.
+                    if let Some(filename) = filename {
+                        format!("{}, filename={}", error, filename)
+                    } else {
+                        error
+                    }
+                }
+                _ => {
+                    format!("{:?}", err)
+                }
+            };
+            let error = CString::new(error).unwrap();
+            let size = std::cmp::min(error.as_bytes_with_nul().len(), SC_CONFIG_ERRBUF_SIZE);
+            std::ptr::copy_nonoverlapping(error.as_ptr(), errbuf, size);
+            return std::ptr::null_mut();
+        }
+    };
+    Box::into_raw(Box::new(config)) as *const _ as *mut _
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ScConfigYamlFree(yaml: *mut Yaml) {
+    let _ = Box::from_raw(yaml);
+}
+
+/// Get, and create if a required a CString for the provided &str.
+fn get_cstring(value: &str) -> *const c_char {
+    let mut cstrings = CSTRINGS.lock().unwrap();
+    if let Some(cs) = cstrings.get(value) {
+        return cs.as_ptr();
+    }
+    let cstring = CString::new(value).unwrap();
+    cstrings.insert(value.to_string(), cstring);
+    cstrings.get(value).unwrap().as_ptr()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ScConfigLoadFromString(
+    input: *const c_char, errptr: *mut *const c_char,
+) -> *mut Yaml {
+    let input = match CStr::from_ptr(input).to_str() {
+        Ok(input) => input,
+        Err(err) => {
+            let error = format!("Failed to convert input to UTF-8: {:?}", err);
+            *errptr = get_cstring(&error);
+            return std::ptr::null_mut();
+        }
+    };
+
+    match crate::loader::load_from_str(input) {
+        Ok(mut config) => Box::into_raw(Box::new(config.pop().unwrap())),
+        Err(err) => {
+            let error = format!("Failed to load config from string: {:?}", err);
+            *errptr = get_cstring(&error);
+            std::ptr::null_mut()
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ScConfGet(key: *const c_char, vptr: *mut *const c_char) -> c_int {
+    let key = match CStr::from_ptr(key).to_str() {
+        Ok(key) => key,
+        Err(_) => return -1,
+    };
+
+    let root = GLOBAL.read().unwrap();
```

Comment:
Does this `GLOBAL` imply we have one single config? No multi-instance? Wonder if we should consider this for the lib and multi-tenancy/multi-instance type scenarios we discussed. This would of course require all the callers to be aware of a config pointer/object, so quite intrusive.

Reply:
Both are possible. This is a place holder for the global configuration that we already have, so the API wouldn't change that much.  Its not really being used right now as I'm converting to the C data structures we have, but if we did convert all the C configuration code to use the APIs provided by the configuration crate, this is where we'd put that global stuff to avoid passing it around, as you said, it would be quite intrusive to get rid of the global state.

One idea could be that the lib exposes no concept of a global configuration, and this is done by Suricata the application, with suitable wrappers around the config api to provide a global state, or work our way out of a global state.

But I think thats all more intrusive than the first goal I'm trying to hit here, which is to get a better picture of the running configuration and hardcoded defaults to the users.